nixops/resources/aws_vpn_connection.py

```diff
@@ -0,0 +1,146 @@
+                                " use --allow-recreate if you want to create a new one".format(self._state['vpnConnectionId']))
+            self.warn("vpn connection definition changed, recreating ...")
+            self._destroy()
+            self._client = None
+
+       self._state['region'] = config['region']
+       self.connect()
+
+       customer_gtw_id = config['customerGatewayId']
+       if customer_gtw_id.startswith("res-"):
+           res = self.depl.get_typed_resource(customer_gtw_id[4:].split(".")[0], "vpc-customer-gateway")
+           customer_gtw_id = res._state['customerGatewayId']
+
+       vpn_gateway_id = config['vpnGatewayId']
+       if vpn_gateway_id.startswith("res-"):
+           res = self.depl.get_typed_resource(vpn_gateway_id[4:].split(".")[0], "aws-vpn-gateway")
+           vpn_gateway_id = res._state['vpnGatewayId']
+
+       self.log("creating vpn connection between customer gateway {0} and vpn gateway {1}".format(customer_gtw_id, vpn_gateway_id))
+       vpn_connection = self._client.create_vpn_connection(
+            CustomerGatewayId=customer_gtw_id,
+            VpnGatewayId=vpn_gateway_id,
+            Type="ipsec.1",
+            Options={
+                'StaticRoutesOnly': config['staticRoutesOnly']
+                }
+            )
+
+       vpn_conn_id = vpn_connection['VpnConnection']['VpnConnectionId']
+       with self.depl._db:
+           self.state = self.UP
+           self._state['vpnConnectionId'] = vpn_conn_id
+           self._state['vpnGatewayId'] = vpn_gateway_id
+           self._state['customerGatewayId'] = customer_gtw_id
+           self._state['staticRoutesOnly'] = config['staticRoutesOnly']
+
+    def realize_update_tag(self, allow_recreate):
+        config = self.get_defn()
+        self.connect()
+        tags = config['tags']
+        tags.update(self.get_common_tags())
+        self._client.create_tags(Resources=[self._state['vpnConnectionId']], Tags=[{"Key": k, "Value": tags[k]} for k in tags])
+
+    def _destroy(self):
+        if self.state == self.UP:
+            self.log("deleting vpn connection {}".format(self._state['vpnConnectionId']))
+            self.connect()
+            try:
+                self._client.delete_vpn_connection(
+                    VpnConnectionId=self._state['vpnConnectionId'])
+            except botocore.exceptions.ClientError as e:
+                if e.response['Error']['Code'] == "InvalidVpnConnectionID.NotFound":
+                    self.warn("vpn connection {} was already deleted".format(self._state['vpnConnectionId']))
+                else:
+                    raise e
+
+        with self.depl._db:
+            self.state = self.MISSING
+            self._state['vpnConnectionId'] = None
+            self._state['vpnGatewayId'] = None
+            self._state['customerGatewayId'] = None
+            self._state['staticRoutesOnly'] = None
```

Comment:
It would be nice if we could eliminate this kind of boilerplate code:

* The check for `self.UP` could be eliminated if we only call `destroy` in a created state. (If `destroy` were to be implemented as a diff handler implementing an UP -> UNKNOWN transition, this would be automatic since `UP` would be a precondition for calling the handler.)

* Unsetting all _state fields could be done automatically after calling `destroy`.